File: frontend/src/models/jobRole.ts
export type JobRoleStatus = 'active' | 'inactive' | 'closed';
import type { Contact } from './contact';

export const JOB_ROLE_STATUS: Record<string, JobRoleStatus> = {
	ACTIVE: 'active',
	INACTIVE: 'inactive',
	CLOSED: 'closed'
};

export interface JobRole {
	id: number;
	public_id: string;
	title: string;
	description?: string;
	status: JobRoleStatus;
	is_visible: boolean;
	no_of_vacancies?: number;
	close_date?: string;
	company_id: number;
	contact_id: number;
	created_by_id?: number;
	location?: {
		cities: string[];
		states?: string[];
		country?: string;
	};
	salary_range?: {
		min?: number;
		max?: number;
		currency?: string;
	};
	experience?: {
		min?: number;
		max?: number;
	};
	requirements?: {
		skills?: string[];
		qualifications?: string[];
		disability_preferred?: string[];
	};
	job_details?: {
		designation?: string;
		workplace_type?: string; // hybrid, onsite, remote
		job_type?: string; // permanent, contract, full time, internship
	};
	other?: any;
	created_at: string;
	updated_at: string;
	company?: {
		id: number;
		public_id: string;
		name: string;
	};
	contact?: Contact;
	creator?: {
		id: number;
		public_id: string;
		full_name?: string;
		username: string;
	};
	mappings_count?: number;
}

export interface JobRoleCreate extends Omit<JobRole, 'id' | 'public_id' | 'created_at' | 'updated_at' | 'company' | 'contact' | 'creator'> { }
export interface JobRoleUpdate extends Partial<JobRoleCreate> { }

export interface JobRolePaginatedResponse {
	items: JobRole[];
	total: number;
}

File: backend/app/schemas/job_role.py
"""Job Role Schemas"""

from typing import Optional, Dict, List, Any
from uuid import UUID
from datetime import datetime, date
from pydantic import BaseModel, ConfigDict, Field
from app.models.job_role import JobRoleStatus


class JobRoleBase(BaseModel):
    title: str = Field(..., max_length=255)
    description: Optional[str] = None
    status: JobRoleStatus = JobRoleStatus.ACTIVE
    is_visible: bool = True
    no_of_vacancies: Optional[int] = None
    close_date: Optional[date] = None
    company_id: int
    contact_id: int
    location: Optional[Dict[str, Any]] = None
    salary_range: Optional[Dict[str, Any]] = None
    experience: Optional[Dict[str, Any]] = None
    requirements: Optional[Dict[str, Any]] = Field(
        None, 
        description='{"skills": [], "qualifications": [], "disability_preferred": []}'
    )
    job_details: Optional[Dict[str, Any]] = None
    other: Optional[Dict[str, Any]] = None


class JobRoleCreate(JobRoleBase):
    pass


class JobRoleUpdate(BaseModel):
    title: Optional[str] = Field(None, max_length=255)
    description: Optional[str] = None
    status: Optional[JobRoleStatus] = None
    is_visible: Optional[bool] = None
    no_of_vacancies: Optional[int] = None
    close_date: Optional[date] = None
    company_id: Optional[int] = None
    contact_id: Optional[int] = None
    location: Optional[Dict[str, Any]] = None
    salary_range: Optional[Dict[str, Any]] = None
    experience: Optional[Dict[str, Any]] = None
    requirements: Optional[Dict[str, Any]] = None
    job_details: Optional[Dict[str, Any]] = None
    other: Optional[Dict[str, Any]] = None


class CompanyMinimal(BaseModel):
    id: int
    public_id: UUID
    name: str

    model_config = ConfigDict(from_attributes=True)


class ContactMinimal(BaseModel):
    id: int
    public_id: UUID
    first_name: str
    last_name: str
    email: Optional[str] = None
    phone: Optional[str] = None
    mobile: Optional[str] = None
    designation: Optional[str] = None
    department: Optional[str] = None

    model_config = ConfigDict(from_attributes=True)


class UserMinimal(BaseModel):
    id: int
    public_id: UUID
    full_name: Optional[str] = None
    username: str

    model_config = ConfigDict(from_attributes=True)


class JobRoleRead(JobRoleBase):
    id: int
    public_id: UUID
    created_by_id: Optional[int] = None
    created_at: datetime
    updated_at: datetime
    company: Optional[CompanyMinimal] = None
    contact: Optional[ContactMinimal] = None
    creator: Optional[UserMinimal] = None
    mappings_count: int = 0

    model_config = ConfigDict(from_attributes=True)


class JobRoleListResponse(BaseModel):
    items: List[JobRoleRead]
    total: int
